src/core/markovProcessGraph.js

```diff
@@ -0,0 +1,605 @@
+      for (const outEdge of outEdges) {
+        const pr = (outEdge.weight / totalOutWeight) * totalOutPr;
+        addEdge({
+          address: outEdge.address,
+          reversed: outEdge.reversed,
+          src: rewriteEpochNode(outEdge.src, outEdge.timestamp),
+          dst: rewriteEpochNode(outEdge.dst, outEdge.timestamp),
+          transitionProbability: pr,
+        });
+      }
+    }
+
+    // Add radiation edges
+    for (const node of _nodes.values()) {
+      if (node.address === SEED_ADDRESS) continue;
+      let type;
+      if (NodeAddress.hasPrefix(node.address, EPOCH_PREFIX)) {
+        type = EPOCH_RADIATION;
+      } else if (NodeAddress.hasPrefix(node.address, CORE_NODE_PREFIX)) {
+        throw new Error(
+          "invariant violation: unknown core node: " +
+            NodeAddress.toString(node.address)
+        );
+      } else {
+        type = CONTRIBUTION_RADIATION;
+      }
+      addEdge({
+        address: EdgeAddress.append(type, ...NodeAddress.toParts(node.address)),
+        reversed: false,
+        src: node.address,
+        dst: SEED_ADDRESS,
+        transitionProbability:
+          1 - NullUtil.orElse(_nodeOutMasses.get(node.address), 0),
+      });
+    }
+
+    return new MarkovProcessGraph(_nodes, _edges, _scoringAddresses);
+  }
+
+  scoringAddresses(): Set<NodeAddressT> {
+    return new Set(this._scoringAddresses);
+  }
+
+  node(address: NodeAddressT): MarkovNode | null {
+    NodeAddress.assertValid(address);
+    return this._nodes.get(address) || null;
+  }
+
+  *nodes(options?: {|+prefix: NodeAddressT|}): Iterator<MarkovNode> {
+    const prefix = options ? options.prefix : NodeAddress.empty;
+    for (const node of this._nodes.values()) {
+      if (NodeAddress.hasPrefix(node.address, prefix)) {
+        yield node;
+      }
+    }
+  }
+
+  *edges(): Iterator<MarkovEdge> {
+    for (const edge of this._edges.values()) {
+      yield edge;
+    }
+  }
+
+  *inNeighbors(nodeAddress: NodeAddressT): Iterator<MarkovEdge> {
+    for (const edge of this._edges.values()) {
+      if (edge.dst !== nodeAddress) {
+        continue;
+      }
+      yield edge;
+    }
+  }
+
+  toMarkovChain(): OrderedSparseMarkovChain {
+    const nodeOrder = Array.from(this._nodes.keys()).sort();
+    const nodeIndex: Map<
+      NodeAddressT,
+      number /* index into nodeOrder */
+    > = new Map();
+    nodeOrder.forEach((n, i) => {
+      nodeIndex.set(n, i);
+    });
+
+    const inNeighbors: Map<NodeAddressT, MarkovEdge[]> = new Map();
+    for (const edge of this._edges.values()) {
+      MapUtil.pushValue(inNeighbors, edge.dst, edge);
+    }
+
+    const chain = nodeOrder.map((addr) => {
+      const inEdges = NullUtil.orElse(inNeighbors.get(addr), []);
+      const inDegree = inEdges.length;
+      const neighbor = new Uint32Array(inDegree);
+      const weight = new Float64Array(inDegree);
+      inEdges.forEach((e, i) => {
+        // Note: We don't group-by src, so there may be multiple `j`
```

Comment:
It would be an efficiency improvement to collapse these together, right? Smaller/denser markov chain that is semantically equivalent. I dont think this happens much in practice so no change requested.